File: adventofcode/2024/2-2.py
# Advent of code Day 2 Part 2

# A list containing lists for all the numbers of each line.
fileList = []
# a text file with the input numbers "reports" with "labs"
file = open("2024/2-1input.txt", "r")
# read in lines from file and break into fileList
for line in file:
    lineList = line.split()
    fileList.append(lineList)
#print(fileList)

# Total number of safe reports
total = 0
# iterate through lines from file as list of numbers.
for i in fileList:
    # the location of the lab to remove
    n = 0
    # additional loop for removing labs
    while n < len(i):
        # this templist needed to be a copy of i so popping would not effect it
        tempList = list(i)
        # remove the lab we want
        tempList.pop(n)
        # while loop counter
        count = 0
        # will turn false if a report is deemed unsafe
        safe = True
        # stores if we are increasing or decreasing
        dir = 0
        # increment n
        n = n + 1
        # loop through each line
        while count < (len(tempList) - 1):
            # our two values to compare
            valueA = int(tempList[count])
            valueB = int(tempList[count + 1])
            # the two values should not match
            if valueA == valueB:
                safe = False
            elif valueA < valueB:
                if dir == 0 or dir == 1:
                    dir = 1
                else:
                    safe = False
                     # make sure the distance is at least one but not more than three
                if valueB - valueA >= 1 and valueB - valueA <=3:
                    pass
                else:
                    safe = False
            elif valueA > valueB:
                if dir == 0 or dir == -1:
                    dir = -1
                else:
                    safe = False
                     # make sure the distance is at least one but not more than three
                if valueA - valueB >= 1 and valueA - valueB <= 3:
                    pass
                else:
                    safe = False
                    


            
            count = count + 1
            #print(safe)
            
            # we don't need to keep testing if a pair is unsafe
            if not safe:
                #print(tempList, "not safe")

                break
        # if a line made it though safe we add to total and break out
        # because we dont need to keep testing removing diff labs
        if safe:
            #print(i, "safe")
            total = total + 1
            break
print("total: " + str(total))

File: adventofcode/2024/2-2.cpp
// Unfinished! I wanted to do this all in c++ to brush up on my skills
// But this one is so much easier for me to do in python
#include <fstream>
#include <iostream>
#include <string>
#include <algorithm>
#include <stdlib.h>
using namespace std;

int main() {
    // Open the text file
    std::ifstream file("2-1input.txt");

    std::string str;

    int count = 0;
    int total = 0;

    while (std::getline(file, str)){
        // cout << str << '\n';
        // count++;

        //20 is arbitrary but should be big enough
        int level[20];
        string number = "";
        bool prespace = true;
        int index = 0;
        //added end of line character, kinda lazy I think
        str = str + "e";
        
        // getting all of the numbers out of the txt and loaded into an array
        for (char i : str){
            // adds an end of line character so I can capture the last value
            if (i == 'e'){
                // convert to integer
                int tnum = stoi(number);
                level[index] = tnum;
                number = "";
                prespace = true;
                
            }
            if (i != ' '){
                // add chars to number until you hit a space
                if (prespace){
                    number = number + i;
                } else {
                    // convert to int
                    int tnum = stoi(number);
                    level[index] = tnum;
                    number = "";
                    number = number + i;
                    index++;
                    prespace = true;
                }
            } else {
                prespace = false;
            }
        }


        // We now have a full "report". We need to check each report if they would still be safe without at least one of their levels.

        // dampened level index. This level will be ignored
        //int dampLev = 0;
        for (int i = 0; i <= index; i++){
            std::cout << i << "\n";


            

            //Testing a instance of a report
            // loop count
            count = 0;
            //says if the report is safe so far
            bool safe = true;
            //direction (1 for positive -1 for negative 0 for nothing)
            int dir = 0;
            int valueA;
            int valueB;
            //loop though levels
            while (count <= index){

                // Dastardly if block incoming
                if (i == count && i == 0){
                    
                    count++;
                    int valueA = level[count];
                    int valueB = level[count + 1];

                } else if (i == count && i == index){
                    
                    break;
                } else if (i == count) {
                    int valueA = level[count - 1];
                    int valueB = level[count + 1];

                    count++;

                }


                if (count > index) {
                    break;
                }

                // needs to increase or decrease
                if (valueA == valueB){
                    cout << valueA << "should not equal" << valueB << "\n";
                    safe = false;
                } else if (valueA < valueB){ // if the values are increasing set direction to 1
                    if (dir == 0 || dir == 1){
                        dir = 1;
                    } else {
                        safe = false;
                        cout << valueA << valueB << "Direction issue \n";
                    }
                    if(valueB - valueA >= 1 && valueB - valueA <= 3){
                        //safe = true;
                        cout << valueA << " " << valueB << "good jump size \n";
                    } else {
                        safe = false;
                        cout << valueA << valueB << "Too big of a jump \n";
                    }
                } else if (valueA > valueB){
                    if (dir == 0 || dir == -1){
                        dir = -1;
                    } else {
                        safe = false;
                        cout << valueA << " " << valueB << "Direction issue \n";
                    }
                    if (valueA - valueB >= 1 && valueA - valueB <=3){
                        cout << valueA << " " << valueB << "good jump size \n";
                        //safe = true;
                    } else {
                        cout << valueA << valueB << "Too big of a jump \n";
                        safe = false;
                    }
                }



                


                // cout << level[count] << "\n";
                if (!safe){
                    cout << str << " unsafe \n";
                    break;
                }
            
            }
            if(safe){
                cout << str << " safe \n";
                total++;
                break;
            }
        }
        
        

        // cout << "NEXT LOOP \n \n \n";


    }
    cout << total << " total \n";

    return 0;
}